preprocess_observation.py

```diff
@@ -21,25 +21,69 @@ def preprocess_observation_helper(mydict):
-            retval.append([index-1, 0])
+            for j in range(7):
+                retval.append([NUM_POKEMON*2 + TEAM_SIZE*2 + NUM_STATUS_CONDITIONS * tempDict[name] + j, 0])
+        # Add whether this Pokemon is active
+        index = pokedex[name.lower()]['num']
+        retval.append([index-1, pokemon['active']])
         retval2[tempDict[name]] = k
         k += 1
     other_indices = [pokedex['houndoom']['num'], pokedex['arceus']['num'], pokedex['cacturne']['num'], pokedex['dragonite']['num'], pokedex['druddigon']['num'], pokedex['uxie']['num']]
     assert(int(mydict['State'][6]) != 6)  # Remove this eventually
     cur_pokemon = int(mydict['State'][6])
-    for i in [0,1,2,3,4,5]:
+    for i in range(6):
         # Append health information
-        retval.append([809*2+6+i, int(mydict['State'][i].split('/')[0].split(' ')[0])])
-        # Append which Pokemon is on the field
-        retval.append([809-1+ other_indices[i], i == cur_pokemon])
+        condition = mydict['State'][i].split('/')[0].split(' ')
+        health = int(condition[0])
+        retval.append([NUM_POKEMON*2 + TEAM_SIZE + i, health])
+        # Add status information
+        if len(condition) != 1 and condition[1] != 'fnt':
+            # in the future, a numerical value (e.g. 2 turns of sleep remaining) would be nice instead of just 1/0.
+            # TODO: confusion is not mutually exclusive with other status conditions, take account of this
+            assert(len(condition) == 2)
+            assert(condition[1] in STATUS_DICT)
+            for j in range(7):
+                retval.append([NUM_POKEMON*2 + TEAM_SIZE*2 + NUM_STATUS_CONDITIONS*TEAM_SIZE + NUM_STATUS_CONDITIONS * i + j, STATUS_DICT[condition[1]] == j])
+        else:
+            for j in range(7):
+                retval.append([NUM_POKEMON*2 + TEAM_SIZE*2 + NUM_STATUS_CONDITIONS*TEAM_SIZE + NUM_STATUS_CONDITIONS * i + j, 0])
 
+        # Append which Pokemon is on the field
+        retval.append([NUM_POKEMON + other_indices[i]-1, i == cur_pokemon])
+    # Deal with stat boosts
+    for i in [7,8,9,10,11,12,13,   14,15,16,17,18,19,20]:
+        retval.append([NUM_POKEMON*2 + TEAM_SIZE*2 + NUM_STATUS_CONDITIONS*TEAM_SIZE*2 + i-7, mydict['State'][i]])
+    # Add weather
+    assert(mydict['State'][21] in WEATHER_DICT)
+    for i in range(6):
+        retval.append([NUM_POKEMON*2 + TEAM_SIZE*2 + NUM_STATUS_CONDITIONS*TEAM_SIZE*2 + NUM_STAT_BOOSTS*2 + i, i == WEATHER_DICT[mydict['State'][21]]])
+    # And terrain
+    assert(mydict['State'][22] in TERRAIN_DICT)
+    for i in range(4):
+        retval.append([NUM_POKEMON*2 + TEAM_SIZE*2 + NUM_STATUS_CONDITIONS*TEAM_SIZE*2 + NUM_STAT_BOOSTS*2 + NUM_WEATHER + i, i == TERRAIN_DICT[mydict['State'][22]]])
+    # And entry hazards!
+    # Hope spikes and toxic spikes don't get confused
+    retval.append([NUM_POKEMON*2 + TEAM_SIZE*2 + NUM_STATUS_CONDITIONS*TEAM_SIZE*2 + NUM_STAT_BOOSTS*2 + NUM_WEATHER + NUM_TERRAIN, 'spikes' in mydict['State'][23]])
```

Comment:
You could make this a bit easier to follow by setting a variable:

    offset = NUM_POKEMON*2 + TEAM_SIZE*2 + 
    NUM_STATUS_CONDITIONS*TEAM_SIZE*2 + NUM_STAT_BOOSTS*2 + 
    NUM_WEATHER + NUM_TERRAIN

and then using it in the subsequent lines.

Reply:
This is a good idea, and I have added it to the top of the todo list.